sharktank/sharktank/utils/tree.py

```diff
@@ -14,59 +15,157 @@
+                (
+                    k,
+                    map_nodes(
+                        v, f, is_leaf, dict_type=dict_type, sequence_type=sequence_type
+                    ),
+                )
+                for k, v in tree.items()
+            )
+        )
     else:
-        return f([map_nodes(v, f, is_leaf) for v in tree])
+        return f(
+            sequence_type(
+                map_nodes(
+                    v, f, is_leaf, dict_type=dict_type, sequence_type=sequence_type
+                )
+                for v in tree
+            )
+        )
 
 
 def map_leaves(
-    tree: Tree, f: Callable[[Tree], Tree], is_leaf: IsLeaf | None = None
+    tree: Tree,
+    f: Callable[[Tree], Tree],
+    is_leaf: IsLeaf = is_leaf_default,
+    *,
+    dict_type: type = dict,
+    sequence_type: type = tuple
 ) -> Tree:
     """Apply `f` for each leaf in the tree."""
-    if is_leaf is None:
-        is_leaf = is_leaf_default
-
-    def f_leaves_and_nodes(t: Tree) -> Tree:
-        if is_leaf(t):
-            return f(t)
-        else:
-            return t
-
-    return map_nodes(tree, f_leaves_and_nodes, is_leaf)
+    if is_leaf(tree):
+        return f(tree)
+    elif isinstance(tree, MappingABC):
+        return dict_type(
+            (
+                k,
+                map_leaves(
+                    v, f, is_leaf, dict_type=dict_type, sequence_type=sequence_type
+                ),
+            )
+            for k, v in tree.items()
+        )
+    else:
+        return sequence_type(
+            map_leaves(v, f, is_leaf, dict_type=dict_type, sequence_type=sequence_type)
+            for v in tree
+        )
 
 
-def flatten(tree: Tree, is_leaf: IsLeaf | None = None) -> Sequence[Leaf]:
+def flatten(tree: Tree, is_leaf: IsLeaf = is_leaf_default) -> Sequence[Leaf]:
     """Get the leaves of the tree."""
-    return [x for x in _flatten(tree, is_leaf)]
+    return [x for x in iterate_leaves(tree, is_leaf)]
 
 
-def _flatten(tree: Tree, is_leaf: IsLeaf | None = None) -> Iterable[Leaf]:
-    if is_leaf is None:
-        is_leaf = is_leaf_default
+def iterate_leaves(tree: Tree, is_leaf: IsLeaf = is_leaf_default) -> Iterable[Leaf]:
     if is_leaf(tree):
         yield tree
     elif isinstance(tree, MappingABC):
         for v in tree.values():
-            yield from flatten(v, is_leaf)
+            yield from iterate_leaves(v, is_leaf)
     else:
         for v in tree:
-            yield from flatten(v, is_leaf)
+            yield from iterate_leaves(v, is_leaf)
+
+
+_initial_missing = object()
+
+
+def reduce_horizontal(
+    fn: Callable[[Any, Leaf], Any],
+    trees: Sequence[Tree],
+    initial: Tree = _initial_missing,
+    is_leaf: IsLeaf = is_leaf_default,
+    dict_type: type = dict,
+    sequence_type: type = tuple,
+) -> Tree:
+    """Reduce a list of trees such that elements with the same path in the tree are
+    aligned.
+
+    E.g.
+    ```
+    {"a": a1, {"b": b1}}
+    {"a": a2, {"b": b2}}
```

Comment:
I don't understand what this example is showing. Is it complete, or what am I missing?